src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -0,0 +1,976 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.List;
+import java.util.function.BiPredicate;
+
+import static java.lang.foreign.ValueLayout.*;
+import static java.lang.invoke.MethodType.methodType;
+import static jdk.internal.org.objectweb.asm.Opcodes.*;
+
+public class BindingSpecializer {
+    private static final String DUMP_CLASSES_DIR
+        = GetPropertyAction.privilegedGetProperty("jdk.internal.foreign.abi.Specializer.DUMP_CLASSES_DIR");
+    private static final boolean PERFORM_VERIFICATION
+        = GetBooleanAction.privilegedGetProperty("jdk.internal.foreign.abi.Specializer.PERFORM_VERIFICATION");
+
+    // Bunch of helper constants
+    private static final int CLASSFILE_VERSION = VM.classFileVersion();
+
+    private static final String OBJECT_DESC = Object.class.descriptorString();
+    private static final String OBJECT_INTRN = Type.getInternalName(Object.class);
+
+    private static final String BINDING_CONTEXT_DESC = Binding.Context.class.descriptorString();
+    private static final String OF_BOUNDED_ALLOCATOR_DESC = methodType(Binding.Context.class, long.class).descriptorString();
+    private static final String OF_SESSION_DESC = methodType(Binding.Context.class).descriptorString();
+    private static final String ALLOCATOR_DESC = methodType(SegmentAllocator.class).descriptorString();
+    private static final String SESSION_DESC = methodType(MemorySession.class).descriptorString();
+    private static final String SESSION_IMPL_DESC = methodType(MemorySessionImpl.class).descriptorString();
+    private static final String CLOSE_DESC = methodType(void.class).descriptorString();
+    private static final String ADDRESS_DESC = methodType(MemoryAddress.class).descriptorString();
+    private static final String COPY_DESC = methodType(void.class, MemorySegment.class, long.class, MemorySegment.class, long.class, long.class).descriptorString();
+    private static final String TO_RAW_LONG_VALUE_DESC = methodType(long.class).descriptorString();
+    private static final String OF_LONG_DESC = methodType(MemoryAddress.class, long.class).descriptorString();
+    private static final String OF_LONG_UNCHECKED_DESC = methodType(MemorySegment.class, long.class, long.class, MemorySession.class).descriptorString();
+    private static final String ALLOCATE_DESC = methodType(MemorySegment.class, long.class, long.class).descriptorString();
+    private static final String HANDLE_UNCAUGHT_EXCEPTION_DESC = methodType(void.class, Throwable.class).descriptorString();
+    private static final String METHOD_HANDLES_INTRN = Type.getInternalName(MethodHandles.class);
+    private static final String CLASS_DATA_DESC = methodType(Object.class, MethodHandles.Lookup.class, String.class, Class.class).descriptorString();
+    private static final String RELEASE0_DESC = methodType(void.class).descriptorString();
+    private static final String ACQUIRE0_DESC = methodType(void.class).descriptorString();
```

Comment:
calling methodType() is quite slow because of the cache of MethodType so maybe those initialization should be done explicitly in a static block to avoid to recompute methodType(long).descriptorString() several times

Reply:
yes, it should be less expensive, the ClassDesc still need to be constructed from Class to allow refactoring.